### PR TITLE
Add permission annotations to tables

### PR DESCRIPTION
**Pull request by komaljai**
@tc_acl will correspond to a 14bit value that encompasses the CRUDXPS permissions for both control and data path. 
So for example, "@tc_acl="CRUS:RXP" 
Here, control path permission is CRUS and data path permission is RXP. This will map to permission value as  0x38a6
https://github.com/p4tc-dev/p4tc-e2e/issues/39

### CODE REVIEW

File: backends/tc/backend.cpp [hunk 1]
```diff
@@ -514,6 +514,109 @@ void ConvertToBackendIR::updateDefaultHitAction(const IR::P4Table *t, IR::TCTabl
...
+            default:
+                ::error(ErrorType::ERR_INVALID,
+                        "tc_acl annotation cannot have '%1%' in access permisson", s);
+        }
+        value |= mask;
+    }
+    return value;
+}
+
+cstring ConvertToBackendIR::HandleTableAccessPermisson(const IR::P4Table *t) {
```
Comment (Sosutha) [inline backends/tc/backend.cpp]: Permissions - spelling to be corrected in many places

File: backends/tc/backend.cpp [hunk 2]
```diff
@@ -514,6 +514,109 @@ void ConvertToBackendIR::updateDefaultHitAction(const IR::P4Table *t, IR::TCTabl
...
+    bool unused_ps[14], IsTableAddOnMiss = false;
+    cstring control_path, data_path;
+    memset(unused_ps, true, 14);
+    for (auto table : add_on_miss_tables) {
+        if (table->name.originalName == t->name.originalName) {
+            IsTableAddOnMiss = true;
+        }
+    }
+    auto find = tablePermissons.find(t->name.originalName);
```
Comment (Sosutha) [inline backends/tc/backend.cpp]: spelling mistake

File: testdata/p4tc_samples_outputs/add_entry_3_errors.p4-stderr [hunk 3]
```diff
@@ -0,0 +1,76 @@
+terminate called after throwing an instance of 'backtrace_exception<std::out_of_range>'
```
Comment (Sosutha) [inline testdata/p4tc_samples_outputs/add_entry_3_errors.p4-stderr]: Please check this file. There is some wrong output

File: backends/tc/backend.cpp [hunk 4]
```diff
@@ -514,6 +514,109 @@ void ConvertToBackendIR::updateDefaultHitAction(const IR::P4Table *t, IR::TCTabl
...
+                                        : DEFAULT_TABLE_CONTROL_PATH_ACCESS;
+    }
+    if (data_path.isNullOrEmpty()) {
+        data_path = IsTableAddOnMiss ? DEFAULT_ADD_ON_MISS_TABLE_DATA_PATH_ACCESS
+                                     : DEFAULT_TABLE_DATA_PATH_ACCESS;
+    }
+
+    if (IsTableAddOnMiss) {
+        auto access = data_path.find('C');
+        if (access == nullptr) {
```
Comment (Sosutha) [inline backends/tc/backend.cpp]: This can be written as if(!access)

**Review by Sosutha — CHANGES_REQUESTED** [review]
https://github.com/p4tc-dev/p4tc-e2e/issues/39 - This expects changes in json files for permissions of tables. Please check and update the code accordingly and address the review comments as well

File: backends/tc/backend.cpp (hunk 1 above)
Comment (komaljai) [reply backends/tc/backend.cpp]: addressed

File: backends/tc/backend.cpp (hunk 2 above)
Comment (komaljai) [reply backends/tc/backend.cpp]: addressed

File: testdata/p4tc_samples_outputs/add_entry_3_errors.p4-stderr (hunk 3 above)
Comment (komaljai) [reply testdata/p4tc_samples_outputs/add_entry_3_errors.p4-stderr]: deleted file

File: backends/tc/backend.cpp (hunk 4 above)
Comment (komaljai) [reply backends/tc/backend.cpp]: addressed

File: backends/tc/introspection.h [hunk 5]
```diff
@@ -97,6 +97,7 @@ struct ActionAttributes {
 
 struct TableAttributes {
     cstring name;
+    cstring permissions;
```
Comment (Sosutha) [inline backends/tc/introspection.h:100]: Please add new field in the constructor as well

File: backends/tc/backend.cpp [hunk 6]
```diff
@@ -521,7 +521,7 @@ void ConvertToBackendIR::updateAddOnMissTable(const IR::P4Table *t) {
             add_on_miss_tables.push_back(t);
             auto tableDefinition = ((IR::TCTable *)table);
             tableDefinition->setTableAddOnMiss();
-            tableDefinition->setTablePermisson(HandleTableAccessPermisson(t));
+            tableDefinition->setTablePermisson(HandleTableAccessPermission(t));
```
Comment (Sosutha) [inline backends/tc/backend.cpp]: setTablePermission - spelling mistake

**Review by Sosutha — CHANGES_REQUESTED** [review]
Please check the comments

**Review by psivanup — APPROVED** [review]
Looks fine

File: backends/tc/introspection.h (hunk 5 above)
Comment (komaljai) [reply backends/tc/introspection.h:100]: addressed

File: backends/tc/backend.cpp (hunk 6 above)
Comment (komaljai) [reply backends/tc/backend.cpp]: addressed

**Review by Sosutha — APPROVED** [review]
Looks fine

**Pull request closed without merging**